### PR TITLE
RC522 fixes

### LINKED ISSUE #1675: RC522_spi scans continiously if no tag is present
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

pip
**ESP (ESP32/ESP8266, Board/Sonoff):**

esp32 dev board
**ESPHome version (latest production, beta, dev branch)**

v1.15.3
**Affected component:**

https://next.esphome.io/components/binary_sensor/rc522.html

**Description of problem:**
Apparently everything works fine:
If the tag is placed on the reader it is recognized and the corresponding binary sensor gets triggered. When its removed, the binary sensor gets unset after a short delay. However, I noticed that the antenna gets powered in each iteration cycle, not only once per second as specified if no tag is present. This results in quite high power consumption ~250 mWatts (without verbose logging over serial which slows down the code), but if a tag is placed on the reader, this drops to acceptable values. 
I analysed the code and found in addition that the binary sensors are not actively turned off, but time out. 
The reader (at least the ones I tested, all on these cheep blue modules) behave a little bit unexpected during operation: If a tag is presented it gets recognized, but the next scan fails with STATUS_COLLITION - which is the same error code I get if no card is detected. 
I propose the attached solution to this: Uid gets an additional property last_seen, that gets set on successful scans. On first failed scan last_seen is unset, and on second failed scan the binary switch gets unset. I attached the implementation. Relevant lines are 185 in the rcc522_spi.h, the  RC522::loop() function lines 114-147 and in RC522BinarySensor::process to actively unset the binary sensor. 

**Problem-relevant YAML-configuration entries:**
```yaml
spi:
  clk_pin: GPIO18
  miso_pin: GPIO19
  mosi_pin: GPIO23

rc522_spi:
  cs_pin: GPIO2
  #reset_pin: GPIO5
  update_interval: 1s
  on_tag:
    then:
      - homeassistant.tag_scanned: !lambda 'return x;'

binary_sensor:
  - platform: rc522_spi
    uid: 80-37-B9-A4
    name: "NFC Card 1"
    id: 'nfc1'
[rc522_spi.cpp.txt](https://github.com/esphome/issues/files/5649638/rc522_spi.cpp.txt)
[rc522_spi.h.txt](https://github.com/esphome/issues/files/5649639/rc522_spi.h.txt)


```

**Logs (if applicable):**

```
PASTE DEBUG LOG HERE

```

**Additional information and things you've tried:**

### LINKED ISSUE #1769: RC522 only works with logging set to verbose
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**
Hass.io

**ESP (ESP32/ESP8266, Board/Sonoff):**

D1 Mini, esp8266

**ESPHome version (latest production, beta, dev branch)**

1.16.0b4 (beta)

**Affected component:**

- RC522 (https://beta.esphome.io/components/binary_sensor/rc522.html)

**Description of problem:**

The reading of tags only works when logging is set to `verbose` for versions 1.16.0b4 and later (1.17.0-dev).

In 1.16.0b3 it was still working when logging was default.

**Problem-relevant YAML-configuration entries:**
```yaml
rc522_spi:
  cs_pin: D8
  update_interval: 1s
  on_tag:
    then:
    - homeassistant.tag_scanned: !lambda 'return x;'
    - if:
        condition:
          switch.is_on: buzzer_enabled
        then:
        - rtttl.play: "success:d=24,o=5,b=100:c,g,b"
    - if:
        condition:
          switch.is_on: led_enabled
        then:
        - light.turn_on:
            id: activity_led
            brightness: 30%
            red: 0%
            green: 100%
            blue: 0%
            flash_length: 500ms
```

**Logs (if applicable):**

With default logging the reading of tags is not working, the logs only show:
```yaml
[23:00:15][D][rc522:121]: Status is now: 1
[23:00:15][D][rc522:121]: Status is now: 3
[23:00:16][D][rc522:121]: Status is now: 1
[23:00:16][D][rc522:121]: Status is now: 3
[23:00:20][D][rc522:121]: Status is now: 1
[23:00:20][D][rc522:121]: Status is now: 3
[23:00:23][D][rc522:121]: Status is now: 1
[23:00:23][D][rc522:121]: Status is now: 3
```

**Additional information and things you've tried:**

### PR DESCRIPTION
**Pull request by glmnet**
## Description:

Converted code to state machine, also removed lots of unused code as we only need the UID.
Should use less memory and less power, should be more reliable as previous algorithm blocked the CPU in a tight loop, this state machine implementation returns from the loop call when the RC522 is busy.

This will only support 4 bytes UID do we need 7 bytes or 10 bytes? I have no cards those type.

Still more cleanup will be needed

**Related issue (if applicable):**
- fixes https://github.com/esphome/issues/issues/1675
- fixes https://github.com/esphome/issues/issues/1769

- closes https://github.com/esphome/esphome/pull/1419

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
I have 7-byte tags (lots of nfc stickers use that type)
Dunno about 10-byte though

**Comment by glmnet** [comment]
> I have 7-byte tags (lots of nfc stickers use that type [type 2])
> 
> Dunno about 10-byte though

Send a PR 🤣

**Comment by jesserockz** [comment]
> > I have 7-byte tags (lots of nfc stickers use that type [type 2])
> > Dunno about 10-byte though
> 
> Send a PR 🤣
> 
> Did it work before? I realized that after I went mad cleaning up all "unused" code.

I have not tried them on the rc522, I will try it out.
And looking at the code you deleted that handles the "more than 4 byte" uids it's quite complex.

**Comment by jesserockz** [comment]
Using this branch

```text
4-byte UID
[17:14:15][D][rc522:196]: Found new tag 'F4-ED-1D-EB'
[17:14:25][D][pn532:159]: Found new tag 'F4-ED-1D-EB'

7-byte uid
[17:14:31][D][pn532:159]: Found new tag '04-0A-22-4A-9B-66-81'
[17:14:36][D][rc522:196]: Found new tag '88-04-0A-22'
```

**Comment by jesserockz** [comment]
on `beta` branch with verbose logging It scans the 7-byte fine.
`[17:23:30][D][rc522:164]: Found new tag '04-0A-22-4A-9B-66-81'`

But 4-bytes via this pr is better than nothing (or something with verbose)

**Comment by glmnet** [comment]
> Using this branch
> 
> ```
> 4-byte UID
> [17:14:15][D][rc522:196]: Found new tag 'F4-ED-1D-EB'
> [17:14:25][D][pn532:159]: Found new tag 'F4-ED-1D-EB'
> 
> 7-byte uid
> [17:14:31][D][pn532:159]: Found new tag '04-0A-22-4A-9B-66-81'
> [17:14:36][D][rc522:196]: Found new tag '88-04-0A-22'
> ```

so this is actually reading the 7 byte UID and then putting some garbage later?

**Comment by jesserockz** [comment]
> > Using this branch
> > ```
> > 4-byte UID
> > [17:14:15][D][rc522:196]: Found new tag 'F4-ED-1D-EB'
> > [17:14:25][D][pn532:159]: Found new tag 'F4-ED-1D-EB'
> > 
> > 7-byte uid
> > [17:14:31][D][pn532:159]: Found new tag '04-0A-22-4A-9B-66-81'
> > [17:14:36][D][rc522:196]: Found new tag '88-04-0A-22'
> > ```
> 
> so this is actually reading the 7 byte UID and then putting some garbage later?

No one line is from pn532 and one line from rc522

**Comment by glmnet** [comment]
Ahh thanks for the clarification. So, it's inaccurate but usable. 
Well, since I don't have the tags it will be kind of shooting bullets in the dark for me, however you can try to do it, if you already get a few bytes you can try to see if you are receiving it full size. The buffer is 9 bytes, you can play around with the buffer sizes, and read data length replies, etc. It might not be necessary to issue more commands to get the full 7 bytes uid.

Otherwise we can state that only 4 bytes are supported.

Thanks

**Comment by rradar** [comment]
@glmnet I can confirm that scanning nfc tags with 7-byte UID now results in a new "style" of reading (before it was the whole 7-byte) like described by @jesserockz when logging is set to `VERBOSE`. When using only the default logging `DEBUG` I am only able to read 4-byte UID. So esphome/issues#1769 looks like it's only "half" fixed with this refactoring :eyes: 

Also there is a kind of breaking change (only within the beta actually and only for 7-byte UIDs) because existing tags get read as new ones now by home assistant:

image

**Comment by jesserockz** [comment]
7-byte uid verbose logs

```text
[12:22:02][V][rc522_spi:035]: read_register_(1a) -> 7
[12:22:02][V][rc522_spi:035]: read_register_(8) -> 4
[12:22:02][V][rc522_spi:035]: read_register_(8) -> 44
[12:22:02][V][rc522_spi:035]: read_register_(8) -> 44
[12:22:02][V][rc522_spi:035]: read_register_(8) -> 44
[12:22:02][V][rc522_spi:035]: read_register_(8) -> 44
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 44
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 45
[12:22:03][V][rc522:308]: pcd_transceive_data_(..., 1, ... ) -> 3
[12:22:03][V][rc522:479]: picc_is_new_card_present_() -> 3
[12:22:03][V][rc522_spi:035]: read_register_(1c) -> 20
[12:22:03][V][rc522:330]: pcd_communicate_with_picc_(12, 48,... 0)
[12:22:03][V][rc522_spi:035]: read_register_(1a) -> 7
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 64
[12:22:03][V][rc522_spi:035]: read_register_(c) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(14) -> 2
[12:22:03][V][rc522_spi:035]: read_register_(18) -> 10
[12:22:03][V][rc522:308]: pcd_transceive_data_(..., 1, ... ) -> 0
[12:22:03][V][rc522:479]: picc_is_new_card_present_() -> 0
[12:22:03][D][rc522:121]: Status is now: 0
[12:22:03][V][rc522:561]: picc_select_(&, 0)
[12:22:03][V][rc522_spi:035]: read_register_(1c) -> 20
[12:22:03][V][rc522:330]: pcd_communicate_with_picc_(12, 48,... 0)
[12:22:03][V][rc522_spi:035]: read_register_(1a) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 64
[12:22:03][V][rc522_spi:035]: read_register_(c) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(14) -> 5
[12:22:03][V][rc522_spi:035]: read_register_(18) -> 10
[12:22:03][V][rc522:306]: pcd_transceive_data_(..., 2,  ) -> 0 [88, 4, 1, 5e, d3]
[12:22:03][V][rc522:432]: pcd_calculate_crc_(..., 7, ...)
[12:22:03][V][rc522_spi:035]: read_register_(a) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(44) -> 28
[12:22:03][V][rc522_spi:035]: read_register_(42) -> 8c
[12:22:03][V][rc522:452]: pcd_calculate_crc_() STATUS_OK
[12:22:03][V][rc522:330]: pcd_communicate_with_picc_(12, 48,... 0)
[12:22:03][V][rc522_spi:035]: read_register_(1a) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 64
[12:22:03][V][rc522_spi:035]: read_register_(c) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(14) -> 3
[12:22:03][V][rc522_spi:035]: read_register_(18) -> 10
[12:22:03][V][rc522:308]: pcd_transceive_data_(..., 9, ... ) -> 0
[12:22:03][V][rc522:432]: pcd_calculate_crc_(..., 1, ...)
[12:22:03][V][rc522_spi:035]: read_register_(a) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(44) -> da
[12:22:03][V][rc522_spi:035]: read_register_(42) -> 17
[12:22:03][V][rc522:452]: pcd_calculate_crc_() STATUS_OK
[12:22:03][V][rc522:330]: pcd_communicate_with_picc_(12, 48,... 0)
[12:22:03][V][rc522_spi:035]: read_register_(1a) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 64
[12:22:03][V][rc522_spi:035]: read_register_(c) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(14) -> 5
[12:22:03][V][rc522_spi:035]: read_register_(18) -> 10
[12:22:03][V][rc522:306]: pcd_transceive_data_(..., 2,  ) -> 0 [4a, 9b, 66, 81, 36]
[12:22:03][V][rc522:432]: pcd_calculate_crc_(..., 7, ...)
[12:22:03][V][rc522_spi:035]: read_register_(a) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(44) -> 34
[12:22:03][V][rc522_spi:035]: read_register_(42) -> 30
[12:22:03][V][rc522:452]: pcd_calculate_crc_() STATUS_OK
[12:22:03][V][rc522:330]: pcd_communicate_with_picc_(12, 48,... 0)
[12:22:03][V][rc522_spi:035]: read_register_(1a) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(8) -> 64
[12:22:03][V][rc522_spi:035]: read_register_(c) -> 0
[12:22:03][V][rc522_spi:035]: read_register_(14) -> 3
[12:22:03][V][rc522_spi:035]: read_register_(18) -> 10
[12:22:03][V][rc522:308]: pcd_transceive_data_(..., 9, ... ) -> 0
[12:22:03][V][rc522:432]: pcd_calculate_crc_(..., 1, ...)
[12:22:03][V][rc522_spi:035]: read_register_(a) -> 4
[12:22:03][V][rc522_spi:035]: read_register_(44) -> fe
[12:22:03][V][rc522_spi:035]: read_register_(42) -> 51
[12:22:03][V][rc522:452]: pcd_calculate_crc_() STATUS_OK
[12:22:03][V][rc522:493]: picc_select_(...) -> 0
[12:22:03][D][rc522:164]: Found new tag '04-01-5E-4A-9B-66-81'
```

**Comment by MatthiasLienhard** [comment]
Is the verbose log confirming 7 byte IDs from this PR, or from the beta branch?

**Comment by jesserockz** [comment]
> Is the verbose log confirming 7 byte IDs from this PR, or from the beta branch?

Neither, that is from the original code as glmnet does not have tags to test so needed to see what it was doing with them to get the uid.

**Comment by MatthiasLienhard** [comment]
I think I have a solution to this, can I make a PR of a PR?

**Comment by jesserockz** [comment]
Yes, you can PR into the `rc522-refactor` branch

**Comment by MatthiasLienhard** [comment]
OK, I'll try tomorrow - I need to tidy it up  - and sleep. But it seems to work. Quite complicated protocol though. Longer Ids have to be selected before the second part can be requested. This involves the CRC calculation.

File: esphome/components/rc522/rc522.cpp [hunk 1]
```diff
@@ -437,9 +443,6 @@ RC522::StatusCode RC522::pcd_calculate_crc_(
   pcd_write_register(COMMAND_REG, PCD_CALC_CRC);    // Start the calculation
 
   // Wait for the CRC calculation to complete. Each iteration of the while-loop takes 17.73μs.
-  // TODO check/modify for other architectures than Arduino Uno 16bit
-
-  // Wait for the CRC calculation to complete. Each iteration of the while-loop takes 17.73us.
   for (uint16_t i = 5000; i > 0; i--) {
```
Comment (glmnet) [inline esphome/components/rc522/rc522.cpp:446]: This still need to be removed and checked in the loop call

**Comment by glmnet** [comment]
@MatthiasLienhard thanks again for the updates, I've made it pass lint checks now. It's working here and @jesserockz checked against his 7 bytes tags too.

Did you check power consumption? I managed to turn off antenna while not scanning and it was working okay before your changes, didn't test it now, do you think it's necessary? 

Also there is now in use the crc function which does the for loop, I believe it should be modified to follow the `awaiting_` thing so wifi, and other components gets CPU time instead of waiting there.

The 1ms delay inline is ok.

ESPHome loop is [16ms](https://github.com/esphome/esphome/blob/bccaa78a90a3cbc2fee644a5b30133ee957f511e/esphome/core/application.h#L245) so the checks will be every 16ms

**Comment by MatthiasLienhard** [comment]
Yes, the loop needs to be refactored, the result needs to be read from a different register though: One byte from 0x21<<1 and one from 0x22<<1, maybe this can be depending on an additional state_?
The antenna is still switched off in STATE_READ_SERIAL_DONE, and switched on in each update() on state STATE_INIT. I still expect power consumption to be slightly higher when a tag is present, as there is more communication with tags.
Maybe there should be an option to switch off polling. 

I experienced crashes of the module during my long term testing, hence I added the automatic reset if there is no response from the reader for 10 consecutive requests (line 389). I am not sure whether this solves the issue, as it did not get triggered yet. Do you also see this?

**Comment by glmnet** [comment]
Removed the loop for crc, this is working with my 4 bytes tags, so if you guys check it I guess it's good to go

**Review by OttoWinter — APPROVED** [review]
LGTM, let me know if you're ready to merge.

**Comment by glmnet** [comment]
I'm ready, was waiting for people to confirm more than 4 bytes readings are ok. Previous uglier code seems to work meanwhile

**Comment by jesserockz** [comment]
Working fine for all of my 7 bytes tags

**Pull request merged**